// @syncfusion/ej2-maps/src/maps/model/constants.js
/**
 * Maps constants doc
 */
/**
 * Specifies maps load event name.
 * @private
 */
export var load = 'load';
/**
 * Specifies maps loaded event name.
 * @private
 */
export var loaded = 'loaded';
/**
 * Specifies maps click event name.
 * @private
 */
export var click = 'click';
/**
 * Specifies maps loaded event name.
 * @private
 */
export var rightClick = 'rightClick';
/**
 * Specifies maps double click event name.
 * @private
 */
export var doubleClick = 'doubleClick';
/**
 * Specifies maps resize event name.
 * @private
 */
export var resize = 'resize';
/**
 * Specifies the map tooltip render event
 */
export var tooltipRender = 'tooltipRender';
/**
 * Specifies the map shapeSelected event
 */
export var shapeSelected = 'shapeSelected';
/**
 * Specifies the map shapeHighlight event
 */
export var shapeHighlight = 'shapeHighlight';
/**
 * Specifies maps mousemove event name.
 * @private
 */
export var mousemove = 'mousemove';
/**
 * Specifies maps mouseup event name.
 * @private
 */
export var mouseup = 'mouseup';
/**
 * Specifies maps mousedown event name.
 * @private
 */
export var mousedown = 'mousedown';
/**
 * Specifies maps layerRendering event name.
 * @private
 */
export var layerRendering = 'layerRendering';
/**
 * Specifies maps shapeRendering event name.
 * @private
 */
export var shapeRendering = 'shapeRendering';
/**
 * Specifies maps markerRendering event name.
 * @private
 */
export var markerRendering = 'markerRendering';
/**
 * Specifies maps clusterRendering event name.
 * @private
 */
export var markerClusterRendering = 'markerClusterRendering';
/**
 * Specifies maps markerClick event name.
 * @private
 */
export var markerClick = 'markerClick';
/**
 * Specifies maps clusterClick event name.
 * @private
 */
export var markerClusterClick = 'markerClusterClick';
/**
 * Specifies maps markerMouseMove event name.
 * @private
 */
export var markerMouseMove = 'markerMouseMove';
/**
 * Specifies maps clusterMouseMove event name.
 * @private
 */
export var markerClusterMouseMove = 'markerClusterMouseMove';
/**
 * Specifies maps dataLabelRendering event name.
 * @private
 */
export var dataLabelRendering = 'dataLabelRendering';
/**
 * Specifies maps bubbleRendering event name.
 * @private
 */
export var bubbleRendering = 'bubbleRendering';
/**
 * Specifies maps bubbleClick event name.
 * @private
 */
export var bubbleClick = 'bubbleClick';
/**
 * Specifies maps bubbleMouseMove event name.
 * @private
 */
export var bubbleMouseMove = 'bubbleMouseMove';
/**
 * Specifies maps animationComplete event name.
 * @private
 */
export var animationComplete = 'animationComplete';
/**
 * Specifies maps legendRendering event name.
 * @private
 */
export var legendRendering = 'legendRendering';
/**
 * Specifies maps annotationRendering event name.
 * @private
 */
export var annotationRendering = 'annotationRendering';
/**
 * Specifies maps itemSelection event name
 * @private
 */
export var itemSelection = 'itemSelection';
/**
 * Specifies maps itemHighlight event name
 */
export var itemHighlight = 'itemHighlight';
/**
 * Specifies maps beforePrint event name
 */
export var beforePrint = 'beforePrint';
/**
 * Specifies the map zoom in event name
 */
export var zoomIn = 'zoomIn';
/**
 * Specifies the map zoom out event name
 */
export var zoomOut = 'zoomOut';
/**
 * Specifies the map pan event name
 */
export var pan = 'pan';
