// @syncfusion/ej2-spreadsheet/src/spreadsheet/actions/resize.js
import { closest, EventHandler, isNullOrUndefined } from '@syncfusion/ej2-base';
import { colWidthChanged, rowHeightChanged, contentLoaded } from '../common/index';
import { setRowHeight } from '../../workbook/base/row';
import { getColumn } from '../../workbook/base/column';
/**
 * The `Resize` module is used to handle the resizing functionalities in Spreadsheet.
 */
var Resize = /** @class */ (function () {
    /**
     * Constructor for resize module in Spreadsheet.
     * @private
     */
    function Resize(parent) {
        this.parent = parent;
        this.addEventListener();
    }
    Resize.prototype.addEventListener = function () {
        this.parent.on(contentLoaded, this.wireEvents, this);
    };
    Resize.prototype.wireEvents = function () {
        var rowHeader = this.parent.getRowHeaderContent();
        var colHeader = this.parent.getColumnHeaderContent();
        EventHandler.add(colHeader, 'dblclick', this.dblClickHandler, this);
        EventHandler.add(rowHeader, 'dblclick', this.dblClickHandler, this);
        EventHandler.add(this.parent.getColumnHeaderContent(), 'mousedown', this.mouseDownHandler, this);
        EventHandler.add(this.parent.getRowHeaderContent(), 'mousedown', this.mouseDownHandler, this);
        this.wireResizeCursorEvent(rowHeader, colHeader);
    };
    Resize.prototype.wireResizeCursorEvent = function (rowHeader, colHeader) {
        EventHandler.add(rowHeader, 'mousemove', this.setTarget, this);
        EventHandler.add(colHeader, 'mousemove', this.setTarget, this);
    };
    Resize.prototype.unWireResizeCursorEvent = function () {
        EventHandler.remove(this.parent.getRowHeaderContent(), 'mousemove', this.setTarget);
        EventHandler.remove(this.parent.getColumnHeaderContent(), 'mousemove', this.setTarget);
    };
    Resize.prototype.unwireEvents = function () {
        EventHandler.remove(this.parent.getColumnHeaderContent(), 'dblclick', this.dblClickHandler);
        EventHandler.remove(this.parent.getRowHeaderContent(), 'dblclick', this.dblClickHandler);
        EventHandler.remove(this.parent.getColumnHeaderContent(), 'mousedown', this.mouseDownHandler);
        EventHandler.remove(this.parent.getRowHeaderContent(), 'mousedown', this.mouseDownHandler);
        this.unWireResizeCursorEvent();
    };
    Resize.prototype.removeEventListener = function () {
        if (!this.parent.isDestroyed) {
            this.parent.off(contentLoaded, this.wireEvents);
        }
    };
    Resize.prototype.mouseMoveHandler = function (e) {
        var sheetPanel = this.parent.element.getElementsByClassName('e-sheet-panel')[0];
        var colResizeHandler = this.parent.element.getElementsByClassName('e-colresize-handler')[0];
        var rowResizeHandler = this.parent.element.getElementsByClassName('e-rowresize-handler')[0];
        if (colResizeHandler || rowResizeHandler) {
            if (colResizeHandler) {
                if (e.x > this.trgtEle.parentElement.firstChild.getBoundingClientRect().left) {
                    colResizeHandler.style.left = e.clientX - this.parent.element.getBoundingClientRect().left + 'px';
                }
            }
            else if (rowResizeHandler) {
                if (e.y >= this.trgtEle.parentElement.parentElement.firstChild.getBoundingClientRect().top) {
                    rowResizeHandler.style.top = e.clientY - sheetPanel.getBoundingClientRect().top + 'px';
                }
            }
        }
    };
    Resize.prototype.mouseDownHandler = function (e) {
        this.event = e;
        this.trgtEle = e.target;
        this.updateTarget(e, this.trgtEle);
        var trgt = this.trgtEle;
        var className = trgt.classList.contains('e-colresize') ? 'e-colresize-handler' :
            trgt.classList.contains('e-rowresize') ? 'e-rowresize-handler' : '';
        this.createResizeHandler(trgt, className);
        this.unWireResizeCursorEvent();
        EventHandler.add(this.parent.element, 'mousemove', this.mouseMoveHandler, this);
        EventHandler.add(document, 'mouseup', this.mouseUpHandler, this);
    };
    Resize.prototype.mouseUpHandler = function (e) {
        var colResizeHandler = this.parent.element.getElementsByClassName('e-colresize-handler')[0];
        var rowResizeHandler = this.parent.element.getElementsByClassName('e-rowresize-handler')[0];
        this.resizeOn(e);
        var resizeHandler = colResizeHandler ? colResizeHandler : rowResizeHandler;
        if (resizeHandler) {
            this.parent.element.getElementsByClassName('e-sheet-panel')[0].removeChild(resizeHandler);
            this.updateCursor(e);
        }
        EventHandler.remove(document, 'mouseup', this.mouseUpHandler);
        EventHandler.remove(this.parent.element, 'mousemove', this.mouseMoveHandler);
        this.wireResizeCursorEvent(this.parent.getRowHeaderContent(), this.parent.getColumnHeaderContent());
    };
    Resize.prototype.dblClickHandler = function (e) {
        this.trgtEle = e.target;
        this.updateTarget(e, this.trgtEle);
        var trgt = this.trgtEle;
        if (trgt.classList.contains('e-colresize') || trgt.classList.contains('e-rowresize')) {
            var colIndx = parseInt(trgt.getAttribute('aria-colindex'), 10) - 1;
            var rowIndx = parseInt(this.trgtEle.parentElement.getAttribute('aria-rowindex'), 10) - 1;
            if (trgt.classList.contains('e-colresize')) {
                this.setAutofit(colIndx, true);
            }
            else if (trgt.classList.contains('e-rowresize')) {
                this.setAutofit(rowIndx, false);
            }
        }
    };
    Resize.prototype.setTarget = function (e) {
        var trgt = e.target;
        var newTrgt;
        var tOffsetV;
        var eOffsetV;
        var tClass;
        if (closest(trgt, '.e-header-row')) {
            eOffsetV = e.offsetX;
            tOffsetV = trgt.offsetWidth;
            tClass = 'e-colresize';
            if (!isNullOrUndefined(trgt.previousElementSibling)) {
                newTrgt = trgt.previousElementSibling;
            }
        }
        else if (closest(trgt, '.e-row')) {
            eOffsetV = e.offsetY;
            tOffsetV = trgt.offsetHeight;
            tClass = 'e-rowresize';
            if (!isNullOrUndefined(trgt.parentElement.previousElementSibling)) {
                newTrgt =
                    trgt.parentElement.previousElementSibling.firstElementChild;
            }
        }
        if (tOffsetV - 2 < 8 && eOffsetV !== Math.ceil((tOffsetV - 2) / 2)) {
            if (eOffsetV < Math.ceil((tOffsetV - 2) / 2)) {
                trgt.classList.add(tClass);
                newTrgt.classList.add(tClass);
            }
            else if (eOffsetV > Math.ceil((tOffsetV - 2) / 2)) {
                trgt.classList.add(tClass);
            }
        }
        else if (tOffsetV - 5 < eOffsetV && eOffsetV <= tOffsetV && tOffsetV >= 10) {
            trgt.classList.add(tClass);
        }
        else if (eOffsetV < 5 && newTrgt && tOffsetV >= 10) {
            trgt.classList.add(tClass);
            newTrgt.classList.add(tClass);
        }
        else {
            var resEle = (tClass === 'e-colresize' ? trgt.parentElement.getElementsByClassName(tClass)
                : this.parent.getRowHeaderTable().getElementsByClassName(tClass));
            for (var index = 0; index < resEle.length; index++) {
                resEle[index].classList.remove(tClass);
            }
        }
    };
    Resize.prototype.updateTarget = function (e, trgt) {
        if (closest(trgt, '.e-header-row')) {
            if ((trgt.offsetWidth < 10 && e.offsetX < Math.ceil((trgt.offsetWidth - 2) / 2)) || (e.offsetX < 5 &&
                trgt.offsetWidth >= 10) && trgt.classList.contains('e-colresize') && trgt.previousElementSibling) {
                this.trgtEle = trgt.previousElementSibling;
            }
        }
        else if (closest(trgt, '.e-row')) {
            if ((trgt.offsetHeight < 10 && e.offsetY < Math.ceil((trgt.offsetHeight - 2) / 2)) || (e.offsetY < 5 &&
                trgt.offsetHeight >= 10) && trgt.classList.contains('e-rowresize') && trgt.parentElement.previousElementSibling) {
                this.trgtEle = trgt.parentElement.previousElementSibling.getElementsByClassName('e-header-cell')[0];
            }
        }
    };
    Resize.prototype.setAutofit = function (idx, isCol) {
        var index;
        var oldIdx = idx;
        if (this.parent.scrollSettings.enableVirtualization) {
            idx = isCol ? idx - this.parent.viewport.leftIndex : idx - this.parent.viewport.topIndex;
        }
        var sheet = this.parent.getActiveSheet();
        var mainContent = this.parent.getMainContent();
        var oldValue = isCol ?
            mainContent.getElementsByTagName('col')[idx].style.width : mainContent.getElementsByTagName('tr')[idx].style.height;
        var headerTable = isCol ? this.parent.getColHeaderTable() : this.parent.getRowHeaderTable();
        var contentRow = mainContent.getElementsByClassName('e-row');
        var contentClone = [];
        var contentTable = mainContent.getElementsByClassName('e-content-table')[0];
        var headerRow = headerTable.getElementsByTagName('tr');
        var headerText;
        if (isCol) {
            headerText = headerRow[0].getElementsByClassName('e-header-cell')[idx].cloneNode(true);
            for (index = 0; index < contentRow.length; index++) {
                contentClone[index] = contentRow[index].getElementsByTagName('td')[idx].cloneNode(true);
            }
        }
        else {
            headerText = headerRow[idx].getElementsByClassName('e-header-cell')[0].cloneNode(true);
            for (index = 0; index < contentRow[idx].getElementsByTagName('td').length; index++) {
                contentClone[index] = contentRow[idx].getElementsByTagName('td')[index].cloneNode(true);
            }
        }
        var headerFit = this.findMaxValue(headerTable, [headerText], isCol);
        var contentFit = this.findMaxValue(contentTable, contentClone, isCol);
        var autofitValue = headerFit < contentFit ? contentFit : headerFit;
        var threshold = parseInt(oldValue, 10) > autofitValue ?
            -(parseInt(oldValue, 10) - autofitValue) : autofitValue - parseInt(oldValue, 10);
        if (isCol) {
            getColumn(sheet, idx).width = autofitValue > 0 ? autofitValue : 0;
            this.parent.notify(colWidthChanged, { threshold: threshold, colIdx: oldIdx });
        }
        else {
            setRowHeight(sheet, idx, autofitValue > 0 ? autofitValue : 0);
            this.parent.notify(rowHeightChanged, { threshold: threshold, rowIdx: oldIdx });
        }
        this.setResize(idx, autofitValue + 'px', isCol);
    };
    Resize.prototype.findMaxValue = function (table, text, isCol) {
        var myTableDiv = this.parent.createElement('div', { className: this.parent.element.className });
        var myTable = this.parent.createElement('table', {
            className: table.className + 'e-resizetable',
            styles: 'width: auto;height: auto'
        });
        var myTr = this.parent.createElement('tr');
        if (isCol) {
            text.forEach(function (element) {
                var tr = myTr.cloneNode();
                tr.appendChild(element);
                myTable.appendChild(tr);
            });
        }
        else {
            text.forEach(function (element) {
                myTr.appendChild(element.cloneNode(true));
            });
            myTable.appendChild(myTr);
        }
        myTableDiv.appendChild(myTable);
        document.body.appendChild(myTableDiv);
        var offsetWidthValue = myTable.getBoundingClientRect().width;
        var offsetHeightValue = myTable.getBoundingClientRect().height;
        document.body.removeChild(myTableDiv);
        if (isCol) {
            return Math.ceil(offsetWidthValue);
        }
        else {
            return Math.ceil(offsetHeightValue);
        }
    };
    Resize.prototype.createResizeHandler = function (trgt, className) {
        var editor = this.parent.createElement('div', { className: className });
        if (trgt.classList.contains('e-colresize')) {
            editor.style.height = this.parent.getMainContent().clientHeight + trgt.offsetHeight + 'px';
            editor.style.left = this.event.clientX - this.parent.element.getBoundingClientRect().left + 'px';
            editor.style.top = '0px';
        }
        else if (trgt.classList.contains('e-rowresize')) {
            editor.style.width = this.parent.getMainContent().clientWidth + trgt.offsetWidth + 'px';
            editor.style.left = '0px';
            editor.style.top = this.event.clientY
                - this.parent.element.getElementsByClassName('e-sheet-panel')[0].getBoundingClientRect().top + 'px';
        }
        this.parent.element.getElementsByClassName('e-sheet-panel')[0].appendChild(editor);
        this.updateCursor(this.event);
    };
    Resize.prototype.setColWidth = function (index, width) {
        var sheet = this.parent.getActiveSheet();
        var eleWidth = parseInt(this.parent.getMainContent().getElementsByTagName('col')[index].style.width, 10);
        var colWidth = width;
        var threshold = parseInt(colWidth, 10) - eleWidth;
        if (threshold < 0 && eleWidth < -(threshold)) {
            threshold = -eleWidth;
        }
        var oldIdx = parseInt(this.trgtEle.getAttribute('aria-colindex'), 10) - 1;
        this.parent.notify(colWidthChanged, { threshold: threshold, colIdx: oldIdx });
        this.setResize(index, colWidth, true);
        getColumn(sheet, index).width = parseInt(colWidth, 10) > 0 ? parseInt(colWidth, 10) : 0;
        sheet.columns[index].customWidth = true;
        this.parent.setProperties({ sheets: this.parent.sheets }, true);
    };
    Resize.prototype.setRowHeight = function (index, height) {
        var eleHeight = parseInt(this.parent.getMainContent().getElementsByTagName('tr')[index].style.height, 10);
        var rowHeight = height;
        var threshold = parseInt(rowHeight, 10) - eleHeight;
        if (threshold < 0 && eleHeight < -(threshold)) {
            threshold = -eleHeight;
        }
        var oldIdx = parseInt(this.trgtEle.parentElement.getAttribute('aria-rowindex'), 10) - 1;
        this.parent.notify(rowHeightChanged, { threshold: threshold, rowIdx: oldIdx });
        this.setResize(index, rowHeight, false);
        setRowHeight(this.parent.getActiveSheet(), index, parseInt(rowHeight, 10) > 0 ? parseInt(rowHeight, 10) : 0);
        this.parent.getActiveSheet().rows[index].customHeight = true;
        this.parent.setProperties({ sheets: this.parent.sheets }, true);
    };
    Resize.prototype.resizeOn = function (e) {
        var idx;
        if (this.trgtEle.classList.contains('e-rowresize')) {
            idx = parseInt(this.trgtEle.parentElement.getAttribute('aria-rowindex'), 10) - 1;
            if (this.parent.scrollSettings.enableVirtualization) {
                idx = idx - this.parent.viewport.topIndex;
            }
            var rowHeight = e.clientY - this.event.clientY +
                parseInt(this.parent.getMainContent().getElementsByClassName('e-row')[idx].style.height, 10) + 'px';
            this.setRowHeight(idx, rowHeight);
        }
        else if (this.trgtEle.classList.contains('e-colresize')) {
            idx = parseInt(this.trgtEle.getAttribute('aria-colindex'), 10) - 1;
            if (this.parent.scrollSettings.enableVirtualization) {
                idx = idx - this.parent.viewport.leftIndex;
            }
            var colWidth = e.clientX - this.event.clientX +
                parseInt(this.parent.getMainContent().getElementsByTagName('col')[idx].style.width, 10) + 'px';
            this.setColWidth(idx, colWidth);
        }
    };
    Resize.prototype.setWidthAndHeight = function (trgt, value, isCol) {
        if (isCol) {
            trgt.style.width = parseInt(trgt.style.width, 10) + value + 'px';
        }
        else {
            trgt.style.height = parseInt(trgt.style.height, 10) + value + 'px';
        }
    };
    // tslint:disable-next-line:max-func-body-length
    Resize.prototype.setResize = function (index, value, isCol) {
        var curEle;
        var curEleH;
        var curEleC;
        var preEle;
        var preEleH;
        var preEleC;
        var nxtEle;
        var nxtEleH;
        var nxtEleC;
        var sheet = this.parent.getActiveSheet();
        if (isCol) {
            curEle = this.parent.element.getElementsByClassName('e-column-header')[0].getElementsByTagName('th')[index];
            curEleH = this.parent.element.getElementsByClassName('e-column-header')[0].getElementsByTagName('col')[index];
            curEleC = this.parent.element.getElementsByClassName('e-main-content')[0].getElementsByTagName('col')[index];
        }
        else {
            curEle = this.parent.element.getElementsByClassName('e-row-header')[0].getElementsByTagName('tr')[index];
            curEleH = this.parent.element.getElementsByClassName('e-row-header')[0].getElementsByTagName('tr')[index];
            curEleC = this.parent.element.getElementsByClassName('e-main-content')[0].getElementsByTagName('tr')[index];
            curEleH.style.height = parseInt(value, 10) > 0 ? value : '2px';
            curEleC.style.height = parseInt(value, 10) > 0 ? value : '0px';
            var hdrRow = this.parent.getRowHeaderContent().getElementsByClassName('e-row');
            var hdrClone = [];
            hdrClone[0] = hdrRow[index].getElementsByTagName('td')[0].cloneNode(true);
            var hdrFntSize = this.findMaxValue(this.parent.getRowHeaderTable(), hdrClone, false) + 1;
            var contentRow = this.parent.getMainContent().getElementsByClassName('e-row');
            var contentClone = [];
            for (var idx = 0; idx < contentRow[index].getElementsByTagName('td').length; idx++) {
                contentClone[idx] = contentRow[index].getElementsByTagName('td')[idx].cloneNode(true);
            }
            var cntFntSize = this.findMaxValue(this.parent.getContentTable(), contentClone, false) + 1;
            var fntSize = hdrFntSize >= cntFntSize ? hdrFntSize : cntFntSize;
            if (parseInt(curEleC.style.height, 10) < fntSize ||
                (curEle.classList.contains('e-reach-fntsize') && parseInt(curEleC.style.height, 10) === fntSize)) {
                curEle.classList.add('e-reach-fntsize');
                curEleH.style.lineHeight = parseInt(value, 10) >= 4 ? ((parseInt(value, 10)) - 4) + 'px' :
                    parseInt(value, 10) > 0 ? ((parseInt(value, 10)) - 1) + 'px' : '0px';
                curEleC.style.lineHeight = parseInt(value, 10) > 0 ? ((parseInt(value, 10)) - 1) + 'px' : '0px';
            }
            else {
                curEleH.style.removeProperty('line-height');
                curEleC.style.removeProperty('line-height');
                if (curEle.classList.contains('e-reach-fntsize')) {
                    curEle.classList.remove('e-reach-fntsize');
                }
            }
        }
        preEle = curEle.previousElementSibling;
        nxtEle = curEle.nextElementSibling;
        if (preEle) {
            preEle = curEle.previousElementSibling;
            preEleH = curEleH.previousElementSibling;
            preEleC = curEleC.previousElementSibling;
        }
        if (nxtEle) {
            nxtEle = curEle.nextElementSibling;
            nxtEleH = curEleH.nextElementSibling;
            nxtEleC = curEleC.nextElementSibling;
        }
        if (parseInt(value, 10) <= 0 && !(curEle.classList.contains('e-zero') || curEle.classList.contains('e-zero-start'))) {
            if (preEle && nxtEle) {
                if (isCol) {
                    curEleH.style.width = '2px';
                    curEleC.style.width = '0px';
                }
                else {
                    curEleH.style.height = '2px';
                    curEleC.style.height = '0px';
                }
                if (preEle.classList.contains('e-zero-start')) {
                    curEle.classList.add('e-zero-start');
                    curEleC.classList.add('e-zero-start');
                }
                else {
                    curEle.classList.add('e-zero');
                    curEleC.classList.add('e-zero');
                }
                if (!nxtEle.classList.contains('e-zero') && !nxtEle.classList.contains('e-zero-last')) {
                    curEle.classList.add('e-zero-last');
                    curEleC.classList.add('e-zero-last');
                }
                if (preEle.classList.contains('e-zero-last')) {
                    preEle.classList.remove('e-zero-last');
                    preEleC.classList.remove('e-zero-last');
                }
                if (preEle.classList.contains('e-zero')) {
                    if (curEle.classList.contains('e-zero-end')) {
                        this.setWidthAndHeight(preEleH, -2, isCol);
                    }
                    else {
                        this.setWidthAndHeight(preEleH, -2, isCol);
                    }
                }
                else {
                    this.setWidthAndHeight(preEleH, -1, isCol);
                }
                if (preEle.classList.contains('e-zero-start')) {
                    this.setWidthAndHeight(curEleH, -1, isCol);
                }
                if (nxtEle.classList.contains('e-zero')) {
                    if (curEle.classList.contains('e-zero-start')) {
                        while (nxtEle) {
                            if (nxtEle.classList.contains('e-zero') && (parseInt(nxtEleH.style.height, 10) !== 0 && !isCol) ||
                                (parseInt(nxtEleH.style.width, 10) !== 0 && isCol)) {
                                if (isCol) {
                                    curEleH.style.width = parseInt(curEleH.style.width, 10) - 1 + 'px';
                                    nxtEleH.style.width = parseInt(nxtEleH.style.width, 10) - 1 + 'px';
                                }
                                else {
                                    curEleH.style.height = parseInt(curEleH.style.height, 10) - 1 + 'px';
                                    nxtEleH.style.height = parseInt(nxtEleH.style.height, 10) - 1 + 'px';
                                }
                                nxtEle.classList.remove('e-zero');
                                nxtEle.classList.add('e-zero-start');
                                break;
                            }
                            else {
                                var nxtIndex = void 0;
                                nxtEle.classList.remove('e-zero');
                                nxtEle.classList.add('e-zero-start');
                                if (isCol) {
                                    nxtIndex = parseInt(nxtEle.getAttribute('aria-colindex'), 10) - 1;
                                    nxtEle = this.parent.getColHeaderTable().getElementsByTagName('th')[nxtIndex + 1];
                                    nxtEleH = this.parent.getColHeaderTable().getElementsByTagName('col')[nxtIndex + 1];
                                }
                                else {
                                    nxtIndex = parseInt(nxtEle.getAttribute('aria-rowindex'), 10) - 1;
                                    nxtEle = this.parent.getRowHeaderTable().getElementsByTagName('tr')[nxtIndex + 1];
                                    nxtEleH = this.parent.getRowHeaderTable().getElementsByTagName('tr')[nxtIndex + 1];
                                }
                            }
                        }
                    }
                    else {
                        this.setWidthAndHeight(curEleH, -2, isCol);
                    }
                }
                else {
                    if (nxtEle.classList.contains('e-zero-end')) {
                        if (isCol) {
                            curEleH.style.width = '0px';
                        }
                        else {
                            curEleH.style.height = '0px';
                        }
                    }
                    else {
                        this.setWidthAndHeight(nxtEleH, -1, isCol);
                    }
                }
            }
            else if (preEle) {
                if (isCol) {
                    curEleH.style.width = '1px';
                    curEleC.style.width = '0px';
                }
                else {
                    curEleH.style.height = '1px';
                    curEleC.style.height = '0px';
                }
                curEle.classList.add('e-zero-end');
                curEleC.classList.add('e-zero-end');
                curEle.classList.add('e-zero-last');
                curEleC.classList.add('e-zero-last');
                if (preEle.classList.contains('e-zero')) {
                    this.setWidthAndHeight(preEleH, -2, isCol);
                }
                else {
                    this.setWidthAndHeight(preEleH, -1, isCol);
                }
            }
            else if (nxtEle) {
                curEle.classList.add('e-zero-start');
                curEleC.classList.add('e-zero-start');
                if (!nxtEle.classList.contains('e-zero')) {
                    curEle.classList.add('e-zero-last');
                    curEleC.classList.add('e-zero-last');
                }
                if (isCol) {
                    curEleH.style.width = '1px';
                    curEleC.style.width = '0px';
                }
                else {
                    curEleH.style.height = '1px';
                    curEleC.style.height = '0px';
                }
                if (nxtEle.classList.contains('e-zero')) {
                    while (nxtEle) {
                        if (nxtEle.classList.contains('e-zero') && (parseInt(nxtEleH.style.width, 10) !== 0
                            && isCol) || (parseInt(nxtEleH.style.height, 10) !== 0 && !isCol)) {
                            if (isCol) {
                                nxtEleH.style.width = parseInt(nxtEleH.style.width, 10) - 1 + 'px';
                                curEleH.style.width = parseInt(curEleH.style.width, 10) - 1 + 'px';
                            }
                            else {
                                nxtEleH.style.height = parseInt(nxtEleH.style.height, 10) - 1 + 'px';
                                curEleH.style.height = parseInt(curEleH.style.height, 10) - 1 + 'px';
                            }
                            nxtEle.classList.add('e-zero-start');
                            nxtEle.classList.remove('e-zero');
                            break;
                        }
                        else {
                            var nxtIndex = void 0;
                            nxtEle.classList.add('e-zero-start');
                            nxtEle.classList.remove('e-zero');
                            if (isCol) {
                                nxtIndex = parseInt(nxtEle.getAttribute('aria-colindex'), 10) - 1;
                                nxtEleH = this.parent.getColHeaderTable().getElementsByTagName('col')[nxtIndex + 1];
                                nxtEle = this.parent.getColHeaderTable().getElementsByTagName('th')[nxtIndex + 1];
                            }
                            else {
                                nxtIndex = parseInt(nxtEle.getAttribute('aria-rowindex'), 10) - 1;
                                nxtEleH = this.parent.getRowHeaderTable().getElementsByTagName('tr')[nxtIndex + 1];
                                nxtEle = this.parent.getRowHeaderTable().getElementsByTagName('tr')[nxtIndex + 1];
                            }
                        }
                    }
                }
                else {
                    this.setWidthAndHeight(nxtEleH, -1, isCol);
                }
            }
        }
        else if (parseInt(value, 10) > 0) {
            if (isCol) {
                curEleH.style.width = value;
                curEleC.style.width = value;
            }
            else {
                curEleH.style.height = value;
                curEleC.style.height = value;
            }
            if (preEle && nxtEle) {
                if (preEle.classList.contains('e-zero')) {
                    if (curEle.classList.contains('e-zero')) {
                        if (isCol) {
                            preEleH.style.width = parseInt(preEleH.style.width, 10) + 2 + 'px';
                            curEleH.style.width = parseInt(curEleH.style.width, 10) - 1 + 'px';
                        }
                        else {
                            preEleH.style.height = parseInt(preEleH.style.height, 10) + 2 + 'px';
                            curEleH.style.height = parseInt(curEleH.style.height, 10) - 1 + 'px';
                        }
                    }
                    else {
                        this.setWidthAndHeight(curEleH, -1, isCol);
                    }
                }
                else {
                    if (curEle.classList.contains('e-zero')) {
                        this.setWidthAndHeight(preEleH, 1, isCol);
                    }
                    else {
                        if (curEle.classList.contains('e-zero-start')) {
                            if (isCol) {
                                preEleH.style.width = parseInt(preEleH.style.width, 10) + 1 + 'px';
                                curEleH.style.width = parseInt(curEleH.style.width, 10) - 1 + 'px';
                            }
                            else {
                                preEleH.style.height = parseInt(preEleH.style.height, 10) + 1 + 'px';
                                curEleH.style.height = parseInt(curEleH.style.height, 10) - 1 + 'px';
                            }
                        }
                    }
                }
                if (nxtEle.classList.contains('e-zero')) {
                    this.setWidthAndHeight(curEleH, -1, isCol);
                }
                else {
                    if (curEle.classList.contains('e-zero') || curEle.classList.contains('e-zero-start')) {
                        this.setWidthAndHeight(nxtEleH, 1, isCol);
                    }
                }
                if (curEle.classList.contains('e-zero')) {
                    curEle.classList.remove('e-zero');
                }
                if (curEle.classList.contains('e-zero-start')) {
                    curEle.classList.remove('e-zero-start');
                }
                if (curEleC.classList.contains('e-zero')) {
                    curEleC.classList.remove('e-zero');
                }
                if (curEleC.classList.contains('e-zero-start')) {
                    curEleC.classList.remove('e-zero-start');
                }
                if (curEle.classList.contains('e-zero-last')) {
                    curEle.classList.remove('e-zero-last');
                }
                if (curEleC.classList.contains('e-zero-last')) {
                    curEleC.classList.remove('e-zero-last');
                }
                if (preEle.classList.contains('e-zero') || preEle.classList.contains('e-zero-start')) {
                    preEle.classList.add('e-zero-last');
                    preEleC.classList.add('e-zero-last');
                }
            }
            else if (preEle) {
                if (preEle.classList.contains('e-zero')) {
                    if (curEle.classList.contains('e-zero')) {
                        if (isCol) {
                            curEleH.style.width = parseInt(curEleH.style.width, 10) - 1 + 'px';
                            preEleH.style.width = parseInt(preEleH.style.width, 10) + 2 + 'px';
                        }
                        else {
                            curEleH.style.height = parseInt(curEleH.style.height, 10) - 1 + 'px';
                            preEleH.style.height = parseInt(preEleH.style.height, 10) + 2 + 'px';
                        }
                    }
                    else {
                        this.setWidthAndHeight(curEleH, -1, isCol);
                    }
                }
                else {
                    if (curEle.classList.contains('e-zero')) {
                        this.setWidthAndHeight(preEleH, 1, isCol);
                    }
                    else {
                        this.setWidthAndHeight(curEleH, -1, isCol);
                    }
                }
                if (curEle.classList.contains('e-zero')) {
                    curEle.classList.remove('e-zero');
                }
                if (curEle.classList.contains('e-zero-end')) {
                    curEle.classList.remove('e-zero-end');
                }
                if (curEleC.classList.contains('e-zero')) {
                    curEleC.classList.remove('e-zero');
                }
                if (curEleC.classList.contains('e-zero-end')) {
                    curEleC.classList.remove('e-zero-end');
                }
            }
            else if (nxtEle) {
                if (nxtEle.classList.contains('e-zero')) {
                    this.setWidthAndHeight(curEleH, -1, isCol);
                }
                else if (curEle.classList.contains('e-zero-start')) {
                    this.setWidthAndHeight(nxtEleH, 1, isCol);
                    curEle.classList.remove('e-zero-start');
                }
                if (curEle.classList.contains('e-zero')) {
                    curEle.classList.remove('e-zero');
                }
                if (curEleC.classList.contains('e-zero')) {
                    curEleC.classList.remove('e-zero');
                }
                if (curEle.classList.contains('e-zero-start')) {
                    curEle.classList.remove('e-zero-start');
                }
                if (curEleC.classList.contains('e-zero-start')) {
                    curEleC.classList.remove('e-zero-start');
                }
            }
        }
    };
    Resize.prototype.updateCursor = function (e) {
        if (this.parent.element.getElementsByClassName('e-colresize-handler')[0]) {
            this.parent.element.classList.add('e-col-resizing');
        }
        else if (this.parent.element.classList.contains('e-col-resizing')) {
            this.parent.element.classList.remove('e-col-resizing');
        }
        if (this.parent.element.getElementsByClassName('e-rowresize-handler')[0]) {
            this.parent.element.classList.add('e-row-resizing');
        }
        else if (this.parent.element.classList.contains('e-row-resizing')) {
            this.parent.element.classList.remove('e-row-resizing');
        }
    };
    /**
     * To destroy the resize module.
     * @return {void}
     */
    Resize.prototype.destroy = function () {
        this.unwireEvents();
        this.removeEventListener();
        this.parent = null;
    };
    /**
     * Get the module name.
     * @returns string
     */
    Resize.prototype.getModuleName = function () {
        return 'resize';
    };
    return Resize;
}());
export { Resize };
